File: presentation/src/main/java/com/example/presentation/ui/viewmodel/MainViewModelFactory.java
package com.example.presentation.ui.viewmodel;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.domain.interactor.IAstronomyPictureInteractor;
import com.example.presentation.utils.scheduler.IBaseSchedulerProvider;

import org.jetbrains.annotations.NotNull;


public class MainViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    @NotNull
    private final IAstronomyPictureInteractor mInteractor;
    @NotNull
    private final IBaseSchedulerProvider mSchedulerProvider;
    private final int mCurrentPositionPageAdapter;

    public MainViewModelFactory(@NotNull IAstronomyPictureInteractor interactor,
                                @NotNull IBaseSchedulerProvider schedulerProvider,
                                int currentPositionPageAdapter) {
        mInteractor = interactor;
        mSchedulerProvider = schedulerProvider;
        mCurrentPositionPageAdapter = currentPositionPageAdapter;
    }

    @SuppressWarnings("unchecked cast")
    @NotNull
    @Override
    public <T extends ViewModel> T create(@NotNull Class<T> modelClass) {
        return (T) new MainViewModel(mInteractor, mSchedulerProvider, mCurrentPositionPageAdapter);
    }
}

File: presentation/src/main/java/com/example/presentation/ui/viewmodel/MainViewModelFactory.kt
package com.example.presentation.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import com.example.domain.interactor.IAstronomyPictureInteractor

@Suppress("UNCHECKED_CAST")
class MainViewModelFactory
/**
 * @constructor interactor                 экземпляр интерактора
 * @constructor schedulerProvider          обертка над планировщиками RxJava
 * @constructor currentPositionPageAdapter текущая позиция PageAdapter'a на экране
 */
(
        private val interactor: IAstronomyPictureInteractor,
        private val currentPositionPageAdapter: Int
) : NewInstanceFactory() {

    override fun <T : ViewModel?> create(modelClass: Class<T>) =
            MainViewModel(interactor, currentPositionPageAdapter) as T

}